## Задание 9.5. Модуль ML-6 (HW-03)

In [81]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
#from sklearn.metrics import mean_absolute_error
from sklearn.feature_selection import RFE
from sklearn.feature_selection import SelectKBest, f_classif

In [82]:
data = pd.read_excel('../data/data_ford_price.xlsx') 
display(data)

price  year  condition  cylinders  odometer title_status transmission  \
0     43900  2016          4          6     43500        clean    automatic   
1     15490  2009          2          8     98131        clean    automatic   
2      2495  2002          2          8    201803        clean    automatic   
3      1300  2000          1          8    170305      rebuilt    automatic   
4     13865  2010          3          8    166062        clean    automatic   
...     ...   ...        ...        ...       ...          ...          ...   
7012  22500  2015          3          6     23500        clean    automatic   
7013   5975  2005          2          8         0        clean    automatic   
7014   9999  2006          3          8    161514        clean    automatic   
7015  10900  2011          2          8    164000        clean    automatic   
7016  18000  2010          2          8    104000        clean    automatic   

     drive       size        lat        long  weather  
0      4wd  full-size  36.471500  -82.483400     59.0  
1      4wd  full-size  40.468826  -74.281734     52.0  
2      4wd  full-size  42.477134  -82.949564     45.0  
3      4wd  full-size  40.764373  -82.349503     49.0  
4      4wd        NaN  49.210949 -123.114720      NaN  
...    ...        ...        ...         ...      ...  
7012   rwd  full-size  32.680700 -117.169800     59.0  
7013   rwd  full-size  38.213303  -85.785762     50.0  
7014   NaN  full-size  37.609783 -120.995406     59.0  
7015   4wd  full-size  43.140600  -93.385000     47.0  
7016   4wd  full-size  37.987200  -84.178900     50.0  

[7017 rows x 12 columns]

### Предобработка данных

In [83]:
data = data[['price','year', 'cylinders', 'odometer', 'lat', 'long', 'weather']]
data.dropna(inplace = True)

y = data['price']
x = data.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

### Отбираем три наиболее значимых признака методом RFE

In [84]:
estimator = LinearRegression()
selector = RFE(estimator, n_features_to_select=3, step=1)
selector = selector.fit(X_train, y_train)
 
columns_rfe = selector.get_feature_names_out()
print(columns_rfe)

['year' 'cylinders' 'lat']


### Отбираем три наиболее значимых признака методом SelectKBest

In [85]:
selector = SelectKBest(f_classif, k=3)
selector.fit(X_train, y_train)
 
columns_sb = selector.get_feature_names_out()
print(columns_sb)

['year' 'cylinders' 'odometer']


### Линейная регрессия, обученная на 3 столбцах, отобранных методом RFE

In [86]:
df1 = data[['year', 'cylinders', 'lat']]
df1.dropna(inplace = True)

y = data['price']
x = df1

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

model_1 = LinearRegression()
model_1.fit(X_train, y_train)
y_predicted = model_1.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 5096.570


C:\Users\Надежда\AppData\Local\Temp\ipykernel_12248\979204814.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1.dropna(inplace = True)


### Линейная регрессия, обученная на 3 столбцах, отобранных методом SelectKBest

In [87]:
df2 = data[['year', 'cylinders', 'odometer']]
df2.dropna(inplace = True)

y = data['price']
x = df2

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

model_2 = LinearRegression()
model_2.fit(X_train, y_train)
y_predicted = model_2.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 4708.946


C:\Users\Надежда\AppData\Local\Temp\ipykernel_12248\759033961.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.dropna(inplace = True)


### Вывод.
### Метрика MAE = 5096.570 для модели линейной регрессии, обученной на 3х столбцах, отобранных методом RFE. MAE = 4708.946 - для модели линейной регрессии, обученной на 3х столбцах, отобранных методом SelectKBest. Так как данная метрика помогает оценить абсолютную ошибку (насколько в среднем число в предсказании разошлось с реальным числом), то чем меньше значение метрики, тем точнее модель. Таким образом, в данном случае, для отбора наиболее значимых признаков предпочтительнее метод SelectKBest. 
